File: distconf/cpp/tests/test_distconf.cpp
#include "../DistConf.hpp"
#include <cassert>
#include <chrono>
#include <iostream>
#include <thread>

using namespace distconf;

void test_lifecycle() {
  std::cout << "Testing Lifecycle..." << std::endl;
  {
    DistConfig cfg("standalone");
    // RAII should handle cleanup
  }
  std::cout << "  Passed." << std::endl;
}

void test_data_operations() {
  std::cout << "Testing Data Operations (Get/Set/FullConfig)..." << std::endl;
  DistConfig cfg("standalone");

  // Set/Get
  cfg.Set("cpp_test", "key1", "val1");
  assert(cfg.Get("cpp_test", "key1") == "val1");

  // Full Config
  std::string full = cfg.GetFullConfig();
  assert(full.find("cpp_test") != std::string::npos);
  assert(full.find("val1") != std::string::npos);

  std::cout << "  Passed." << std::endl;
}

void test_capabilities() {
  std::cout << "Testing Capabilities..." << std::endl;
  DistConfig cfg("standalone");

  // Standalone generates default capabilities (config_server, log_server)
  std::string cap = cfg.GetCapability("config_server");
  assert(cap.find("ip") != std::string::npos);

  // Networking
  std::string addr = cfg.GetAddress("config_server");
  assert(!addr.empty());
  assert(addr.find(":") != std::string::npos);

  std::cout << "  Passed." << std::endl;
}

void test_ecosystem_sharing() {
  std::cout << "Testing Ecosystem Sharing (ShareConfig)..." << std::endl;
  DistConfig cfg("standalone");

  // ShareConfig merges top-level keys as sections into LiveConfig
  std::string payload =
      "{\"cpp_service\": {\"status\":\"online\",\"version\":\"2.0\"}}";
  bool success = cfg.ShareConfig(payload);
  assert(success == true);

  // After ShareConfig, data is accessible via Get(section, key)
  std::string status = cfg.Get("cpp_service", "status");
  assert(status == "online");

  std::cout << "  Passed." << std::endl;
}

void test_security() {
  std::cout << "Testing Security (Decrypt)..." << std::endl;
  DistConfig cfg("standalone");

  // ENC(...) detection
  std::string ciphertext = "ENC(dummy)";
  try {
    std::string decrypted = cfg.Decrypt(ciphertext);
    assert(!decrypted.empty());
  } catch (const std::exception &e) {
    // Without keys in CI/CD, we expect a transparent error message
    std::string err = e.what();
    assert(!err.empty());
    std::cout << "  Passed (Caught Transparent Error: " << err << ")"
              << std::endl;
  }

  std::cout << "  Passed." << std::endl;
}

void test_validation() {
  std::cout << "Testing Validation (Fail-Fast)..." << std::endl;
  DistConfig cfg("standalone");

  // Standalone should be valid by default
  assert(cfg.ValidateMandatoryServices() == true);

  std::cout << "  Passed." << std::endl;
}

int main() {
  try {
    test_lifecycle();
    test_data_operations();
    test_capabilities();
    test_ecosystem_sharing();
    test_security();
    test_validation();

    std::cout << "\n=======================================" << std::endl;
    std::cout << "  All C++ SDK Parity Tests Passed! " << std::endl;
    std::cout << "=======================================" << std::endl;
  } catch (const std::exception &e) {
    std::cerr << "\n!!! TEST FAILED: " << e.what() << std::endl;
    return 1;
  }
  return 0;
}

File: distconf/python/tests/test_distconf.py
#!/usr/bin/env python
# coding:utf-8

import unittest
from os import getenv as osGetenv
from os.path import abspath as osPathAbspath, dirname as osPathDirname, exists as osPathExists
from sys import path as sysPath
from time import sleep as timeSleep

# Add parent dir to path to import distconf
sysPath.append(osPathDirname(osPathDirname(osPathAbspath(__file__))))

from distconf import DistConfig

class TestDistConfigFull(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        # Respect LIBDISTCONF_PATH if set, otherwise fallback to relative path
        cls.lib_path = osGetenv("LIBDISTCONF_PATH")
        if not cls.lib_path:
            import platform
            system = platform.system()
            if system == "Darwin":
                ext = ".dylib"
            elif system == "Windows":
                ext = ".dll"
            else:
                ext = ".so"
            cls.lib_path = osPathAbspath(f"../../distconf/libdistconf/libdistconf{ext}")

    # -----------------------------------------------------------------------------------------------

    def test_01_lifecycle(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        self.assertIsNotNone(cfg._handle)
        cfg.close()
        self.assertIsNone(cfg._handle)

    # -----------------------------------------------------------------------------------------------

    def test_02_data_access(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        # Basic Set/Get
        cfg.set("py_test", "key1", "value1")
        self.assertEqual(cfg.get("py_test", "key1"), "value1")
        
        # Get Full Config
        full = cfg.get_full_config()
        self.assertIn("py_test", full.get("live", {}))
        self.assertEqual(full["live"]["py_test"]["key1"], "value1")
        
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_03_capabilities(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        # Resolve address
        addr = cfg.get_address("config_server")
        self.assertTrue(":" in addr)
        
        # Get specific capability
        cap = cfg.get_capability("config_server")
        self.assertIn("ip", cap)
        self.assertIn("port", cap)
        
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_04_sharing(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        payload = {"py_service": {"status": "ok", "items": "1,2,3"}}
        success = cfg.share_config(payload)
        self.assertTrue(success)
        
        # Verify it reflected in LiveConfig
        status = cfg.get("py_service", "status")
        self.assertEqual(status, "ok")
        
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_05_security(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        ciphertext = "ENC(hello)"
        decrypted = cfg.decrypt(ciphertext)
        self.assertIsNotNone(decrypted)
        
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_06_validation(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        self.assertTrue(cfg.validate_mandatory_services())
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_07_callbacks(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        updated_data = []
        def on_update(data):
            updated_data.append(data)
            
        cfg.on_live_conf_update(on_update)
        
        # Trigger update via local set
        cfg.set("callback_test", "trigger", "now")
        
        # Wait for callback dispatch (Go -> C -> Python)
        timeSleep(0.1)
        
        self.assertTrue(len(updated_data) > 0)
        self.assertIn("callback_test", updated_data[0])
        
        cfg.close()

    # -----------------------------------------------------------------------------------------------

    def test_08_registry_callbacks(self):
        cfg = DistConfig("standalone", lib_path=self.lib_path)
        
        registry_events = []
        def on_registry(data):
            registry_events.append(data)
            
        cfg.on_registry_update(on_registry)
        
        # Test that we can register it without crashing. 
        # (Actually triggering it requires injecting network packets or using internal Go hooks, 
        # but registering it ensures the CGO bridge bindings and ctypes wrappers are fully functional).
        self.assertEqual(len(registry_events), 0)
        
        cfg.close()

if __name__ == "__main__":
    unittest.main()
